c10/util/order_preserving_flat_hash_map.h

```diff
@@ -858,24 +959,37 @@ class sherwood_v3_table : private EntryAlloc, private Hasher, private Equal
         {
             current_entry->emplace(distance_from_desired, std::forward<Key>(key), std::forward<Args>(args)...);
             ++num_elements;
+            append_to_list(current_entry);
             return { { current_entry }, true };
         }
         value_type to_insert(std::forward<Key>(key), std::forward<Args>(args)...);
         swap(distance_from_desired, current_entry->distance_from_desired);
+        // We maintain the invariant that:
+        // - result.current_entry contains the new value we're inserting
+        //   and is in the LinkedList position of to_insert
+        // - to_insert contains the value that reprseents the position of
+        //   result.current_entry
         swap(to_insert, current_entry->value);
         iterator result = { current_entry };
         for (++distance_from_desired, ++current_entry;; ++current_entry)
         {
             if (current_entry->is_empty())
             {
                 current_entry->emplace(distance_from_desired, std::move(to_insert));
+                append_to_list(current_entry);
+                // now we can swap back the displaced value to its correct position,
+                // putting the new value we're inserting to the front of the list
+                swap_positions(current_entry, result.current);
                 ++num_elements;
                 return { result, true };
             }
             else if (current_entry->distance_from_desired < distance_from_desired)
             {
                 swap(distance_from_desired, current_entry->distance_from_desired);
                 swap(to_insert, current_entry->value);
+                // to maintain our invariants we need to swap positions
```

Comment:
I think this if condition is the "robinhood hashing" behavior - we could remove this and it would result in less memory locations and the hash map would still work.